### ISSUE TITLE
Allow historical logs import, bypass "entry too far behind"

### ISSUE THREAD

**Issue by vgavro** [grafana/loki#4914]
We have devices that happens to be offline, and may send logs after two weeks for example - producing "entry too far behind" error. I believe that keep "max_chunk_age" too large may cause problems with current logs processing speed (is it? what are caveats?).

It will be great if LOKI will allow to import historical logs not only for current chunk, even having "high cost" for writing such data - it's better then not having this logs at all.

Only alternative I see for now is having two datasources - one database for *current logs* (LOKI), one database for *historical logs* (elasticsearch/kibana?) - which seems excessive because LOKI goals is also log exploration (not only aggregation).

**Comment by vgavro** [comment]
Related to https://github.com/cortexproject/cortex/issues/2366

**Comment by DylanGuedes** [comment]
Hmmm not sure if that's what you are looking for, but another viable option is having two Loki instances, one for online log processing and another one for historical logs, with a much bigger limits (i.e: max_chunk_age). 

Besides the bypass, I'm not sure if Loki can just "bypass" the "entry too far behind" error because a lot of other parameters (ex: chunk size) are used to determine chunkifying behavior and finally you have a compactor compacting these things together in realtime.

**Comment by vgavro** [comment]
@DylanGuedes thanks for reply! I may not know some loki specifics, but like I understand it's using cortex as backend, and there is two relevant approaches that cortex team is considered:
* backfilling https://github.com/cortexproject/cortex/pull/3025
* OR creating new block for too old data - https://github.com/prometheus/prometheus/issues/9607 -  looks like they are willing to choose second approach - See https://github.com/cortexproject/cortex/pull/3025#issuecomment-786027793

Basically, my question is - is LOKI team is willing on working on this issue (and if so - what are prerequisites - is cortex features required for this or not?)

**Comment by DylanGuedes** [comment]
Wow that's very interesting and I didn't know that, thank you!

I can't speak for the Loki team :joy: but from my side, I would be willing to work on it for sure, although I expect it to take quite some time. Anyway, I'll bring this to one to discussion and I'll try to keep this up-to-date.

**Comment by atlee19** [comment]
I also have a requirement to import historical logs and I keep getting the same error "entry too far behind". I'm leaving this comment to agree with @vgavro that having this feature would be useful even though it could result in a higher "cost".

**Comment by DylanGuedes** [comment]
Sorry for taking so long to give you a follow-up on this. You can proceed to import historical data by just disabling `reject_old_samples` and then importing your data will work (as long as it doesn't contain streams that collide with streams in the ingesters). If you are importing streams that contain such collisions, you could try adding a new label to them (such as `job=batch`) just to make them be evaluated as a new stream. In that case, you won't experience `entry too far behind` anymore without the need to tweak `max_chunk_age`.

**Comment by vgavro** [comment]
> Sorry for taking so long to give you a follow-up on this. You can proceed to import historical data by just disabling `reject_old_samples`; then importing your data will work, **as long as it doesn't contain streams that collide with streams in the ingesters**.
> 
> If you are importing streams that contain such collisions, you could try adding a new label to them (such as `job=batch`) just to make the newly imported data to be evaluated as a new stream. In that case, you won't experience `entry too far behind` anymore without the need to tweak `max_chunk_age`.
> 
> edit: if that doesn't make sense/sounds too confusing let me know.

Looks like this will completely solve out problem! Thanks for your reply (didn't looked into source code that well) - really different tags variations are processed in different streams ([docs](https://grafana.com/docs/loki/latest/best-practices/#use-dynamic-labels-sparingly)) , and limitation is per-stream ([code](https://github.com/grafana/loki/blob/v2.4.2/pkg/ingester/stream.go#L259)).
Like I understand it's easy to achieve this way - each DEVICE has it's own dynamic label, and all we have to do is keep sending logs in order per device - that's completely achievable!

**Comment by daveram** [comment]
Just ran into this trying to import previous logs from a few systems using promtail.  Response from loki is "entry too far behind" but we already have reject_old_samples set to false.  Anyone else had it where this doesn't resolve the issue?

Ensured we have a new job label as mentioned above, even restarted loki a few times with a massive reject_old_samples_max_age just in case it was still considering it, but haven't been able to bypass the error yet

**Comment by irizzant** [comment]
Hi
I'm hitting the same issue but the proposed solution:
> Here is my setting : `
reject_old_samples: false
reject_old_samples_max_age: 43800h

schema_config:
configs:
# To be able to store 5 years of data
- from: 2015-01-01
store: boltdb-shipper
object_store: aws
schema: v11
index:
prefix: loki_index_
period: 24h

does not work for me.
I'm adding a simple reproducer using docker-compose [reproducer.zip](https://github.com/grafana/loki/files/8589025/reproducer.zip)

Start the reproducer with `docker-compose up` and then you can try sending this test log:
```
echo '2022-04-27T15:44:31.347267556Z stdout F test' | netcat localhost 5170
```
This is the error loki logs:
```
loki_1_4b79a52b40c1 | level=warn ts=2022-04-29T07:44:31.387262709Z caller=grpc_logging.go:38 method=/logproto.Pusher/Push duration=109.089µs err="rpc error: code = Code(400) desc = entry with timestamp 2022-04-27 15:44:31.347267556 +0000 UTC ignored, reason: 'entry too far behind' for stream: {fluentd_thread=\"flush_thread_0\"},\ntotal ignored: 1 out of 1" msg="gRPC\n"
fluentd_1_cdb20cae605b | 2022-04-29 07:44:31 +0000 [warn]: #0 failed to write post to http://loki:3100/loki/api/v1/push (400 Bad Request entry with timestamp 2022-04-27 15:44:31.347267556 +0000 UTC ignored, reason: 'entry too far behind' for stream: {fluentd_thread="flush_thread_0"},
fluentd_1_cdb20cae605b | total ignored: 1 out of 1
fluentd_1_cdb20cae605b | )
```

**Comment by vacuityv** [comment]
@irizzant  hello, did you solve it ? I met the same problem

**Comment by irizzant** [comment]
@vacuityv no, the [suggested](https://github.com/grafana/loki/issues/4914#issuecomment-1023694138) workaround didn't work for me

**Comment by vacuityv** [comment]
@irizzant  OK，ths. I think maybe it's a timezone problem, but I am not sure. Just wait someone save me.

**Comment by irizzant** [comment]
@vacuityv if you have a look at the reproducer I posted you can see it should be able to ingest samples of years ago, so I don't think it's timezone related

**Comment by irizzant** [comment]
@DylanGuedes thanks for spotting the configuration issue in the compose file, I was using files in my /tmp when I created the reproducer and I oversaw changing the compose file accordingly. 
I may be wrong but I'm quite sure that before it wasn't working though.

**Comment by DylanGuedes** [comment]
No, the `reject_old_samples` is intended to limit data that is old but **is in order**. If you're sending data out of order you're being limited by the max_chunk_age:
```
The validity window for unordered writes is the highest timestamp present minus 1/2 * max-chunk-age.
```

Could you check your max_chunk_age?

**Comment by irizzant** [comment]
I didn't change it in the reproducer so it's the default, but according to the setting reported in [this](https://github.com/grafana/loki/issues/4914#issuecomment-1096667078) comment it was possible to ingest up to 5 years of log in the past so I thought there was no need for tuning `max_chunk_age`.

Ok so if I understand correctly you're saying that if I start from a clean loki without logs stored I can send a log which is 5 years old but if I've already sent some logs to it I have to take into account the `max-chunk-age` right?

**Comment by DylanGuedes** [comment]
> Ok so if I understand correctly you're saying that if I start from a clean loki without logs stored I can send a log which is 5 years old but if I've already sent some logs to it I have to take into account the `max-chunk-age` right?

Yes and no. Yes, in a clean Loki you would be able to send a log which is 5 years old.
If you've already sent some logs **you still can send them**, but the new logs have to have a timestamp that follows the rule of highest timestamp present minus 1/2 * max-chunk-age.

What I'm trying to say is that the problem you're experiencing is more related to your logs being out-of-order than then being many years old.

**Comment by irizzant** [comment]
Ok but I have no control on the order the logs get into loki, so I guess that the only way to import old data is to tweak max-chunk-age.

**Comment by DylanGuedes** [comment]
> Ok but I have no control on the order the logs get into loki, so I guess that the only way to import old data is to tweak max-chunk-age.

Another alternative is sending logs that are too old in a different stream. You shouldn't have **all** your logs with the same set of labels, because this will cause all of them to be evaluated as the same stream.

I'm closing this issue as looks like the issue author was satisfied with the proposed solution. If you find a problem or have a suggestion please open a new one.
